utils/cache_ctx.go

```diff
@@ -0,0 +1,84 @@
+package utils
+
+// Borrowed from Osmosis
+// https://github.com/osmosis-labs/osmosis/blob/62757d309957fa9e02e6fb0b5dc8caf1ca68e696/osmoutils/cache_ctx.go
+// Wraps the execution of a function with a temporary context so that state changes can be discarded if an error occurs
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+	"runtime/debug"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// This function lets you run the function f, but if theres an error or panic
+// drop the state machine change and log the error.
+// If there is no error, proceeds as normal (but with some slowdown due to SDK store weirdness)
+// Try to avoid usage of iterators in f.
+//
+// If its an out of gas panic, this function will also panic like in normal tx execution flow.
+// This is still safe for beginblock / endblock code though, as they do not have out of gas panics.
+func ApplyFuncIfNoError(ctx sdk.Context, f func(ctx sdk.Context) error) (err error) {
```

Comment:
Good question! Considering how small it was, I'd prefer to just pull out the code instead of adding a new dependency. It's also simple enough that we should treat it as our own code instead of treating it as a black box package that we're installing

Reply:
Agreed with pulling the code in for the reasons you listed Sam. We can check in periodically to see if they've made alternations/improvements to the original implementation.